app/controllers/basespace_controller.rb

```diff
@@ -0,0 +1,115 @@
+class BasespaceController < ApplicationController
+  include HttpHelper
+
+  before_action :admin_required
+
+  def oauth
+    @access_token = nil
+
+    if params[:code] &&
+       ENV["BASESPACE_OAUTH_REDIRECT_URI"] &&
+       ENV["BASESPACE_CLIENT_ID"] &&
+       ENV["BASESPACE_CLIENT_SECRET"]
+      response = HttpHelper.post(
+        "https://api.basespace.illumina.com/v1pre3/oauthv2/token",
+        "code" => params[:code],
+        "redirect_uri" => ENV["BASESPACE_OAUTH_REDIRECT_URI"],
+        "client_id" => ENV["BASESPACE_CLIENT_ID"],
+        "client_secret" => ENV["BASESPACE_CLIENT_SECRET"],
+        "grant_type" => "authorization_code"
+      )
+
+      if response.present?
+        @access_token = response["access_token"]
+      end
+    end
+  end
+
+  def projects
+    access_token = params[:access_token]
+
+    if access_token.nil?
+      render json: {
+        error: "basespace access token required"
+      }
+      return
+    end
+
+    # 1024 is the maximum limit allowed by Basespace.
+    # If users reach this limit, we will need to implement multiple requests to fetch all the projects.
+    response = HttpHelper.get(
+      "https://api.basespace.illumina.com/v1pre3/users/current/projects",
+      { limit: 1024 },
+      "Authorization" => "Bearer #{access_token}"
+    )
+
+    # Just return selected fields.
+    formatted_projects = response["Response"]["Items"].map do |dataset|
+      {
+        id: dataset["Id"],
+        name: dataset["Name"]
+      }
+    end
+
+    render json: formatted_projects
+  end
+
+  def samples_for_project
+    access_token = params[:access_token]
+    basespace_project_id = params[:basespace_project_id]
+
+    if access_token.nil?
+      render json: {
+        error: "basespace access token required"
+      }
+      return
+    end
+
+    if basespace_project_id.nil?
+      render json: {
+        error: "basespace project id required"
+      }
+      return
+    end
+
+    # 1024 is the maximum limit allowed by Basespace.
+    # If users reach this limit, we will need to implement multiple requests to fetch all the samples.
+    response = HttpHelper.get(
+      "https://api.basespace.illumina.com/v2/projects/#{basespace_project_id}/datasets?limit=1024",
+      { limit: 1024 },
+      "Authorization" => "Bearer #{access_token}"
+    )
+
+    # Just return selected fields.
+    formatted_samples = response["Items"].map do |dataset|
```

Comment:
Shouldnt we check if response is good before accessing attributes on it? 
I see from your test that in case of errors response might be nil...